Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx

```diff
@@ -21,15 +33,69 @@ const createMarkup = (txt: string) => {
 };
 
 export const RuntimeOutputLog: React.FC<{ projectId: string }> = ({ projectId }) => {
-  const runtimeLogs = useRecoilValue(botRuntimeLogState(projectId));
-  const botRuntimeErrors = useRecoilValue(botRuntimeErrorState(projectId));
+  const runtimeData = useRecoilValue(runtimeStandardOutputDataState(projectId));
+  const botBuildErrors = useRecoilValue(botBuildTimeErrorState(projectId));
+  const { setRuntimeStandardOutputData, setApplicationLevelError } = useRecoilValue(dispatcherState);
+
   const runtimeLogsContainerRef = useRef<HTMLDivElement | null>(null);
 
+  const runtimeTrafficChannel = useRef<WebSocket | null>(null);
+
   useEffect(() => {
     if (runtimeLogsContainerRef?.current) {
       runtimeLogsContainerRef.current.scrollTop = runtimeLogsContainerRef.current.scrollHeight;
     }
-  }, [runtimeLogs, botRuntimeErrors]);
+  }, [runtimeData]);
+
+  useEffect(() => {
+    const setupLogConnection = async () => {
+      try {
+        const runtimeStreamUrl = await httpClient.get(`/publish/runtimeLogUrl/${projectId}`);
+
+        runtimeTrafficChannel.current = new WebSocket(runtimeStreamUrl.data);
+
+        if (runtimeTrafficChannel.current) {
+          runtimeTrafficChannel.current.onmessage = (event) => {
+            const data: { standardError: string; standardOutput: string } = JSON.parse(event.data);
+
+            let standardError: BotStartError | null = null;
+            if (data.standardError) {
+              const isDotnetError = checkIfDotnetVersionMissing({
+                message: data.standardError ?? '',
+              });
+
+              if (isDotnetError) {
+                standardError = {
+                  title: Text.DOTNETFAILURE,
+                  ...missingDotnetVersionError,
+                };
+              } else {
+                standardError = {
+                  title: Text.BOTRUNTIMEERROR,
+                  message: data.standardError,
+                };
+              }
+            }
+            setRuntimeStandardOutputData(projectId, {
+              standardError,
+              standardOutput: data.standardOutput,
+            });
+          };
+        }
+      } catch (ex) {
+        setApplicationLevelError(genericErrorMessage());
+      }
+    };
+
+    if (!runtimeTrafficChannel.current) {
+      setupLogConnection();
+    }
+
+    return () => {
+      runtimeTrafficChannel.current?.close();
+      runtimeTrafficChannel.current = null;
+    };
+  }, []);
```

Comment:
should this function list `projectId` as a dependency? What happens if you open a second project? Will it have only opened the first, initial socket for project A even though you are now using project B?

Reply:
So, i would have added if the component destroy wasn't called on project change. Essentially, when we change project the component is destroyed and the socket as well.